### ISSUE TITLE
Implement fast polynomial evaluation optimizations

### ISSUE THREAD

**Issue by brandonwillard** [pymc-devs/Theano-PyMC#275]
@aseyboldt noticed that we might be missing some (scalar) polynomial optimizations like [Horner's method](https://en.wikipedia.org/wiki/Horner%27s_method).

**Comment by brandonwillard** [comment]
At the very least, we could create an `Op` for NumPy's [`polyval`](https://numpy.org/doc/stable/reference/generated/numpy.polynomial.polynomial.polyval.html) and follow that up with some rewrites for specific cases (e.g. if `x` is a graph input and has a polynomial form in a subgraph, consider replacing that subgraph with a `polyval` `Op`).

**Comment by ricardoV94** [comment]
@brandonwillard is this still up to grabs? I might explore this issue to get a better understanding of Aesara Ops and graph rewrites.

Wrapping a polyval Op around Numpy version should be straightforward.
I have to think a bit about the grad method, but shouldn't be too complicated since we are working with polynomials.

I would definitely need some pointers for the pattern matching / rewriting parts. Let me know if this is a blocker, in which case I am happy to leave this issue to someone more capable :)

**Comment by brandonwillard** [comment]
> is this still up to grabs?

As long as no one is assigned, it should be; however, this particular addition requires some high-level considerations/research first.

For instance, it might be better to add support for generic polynomials backed by `numpy.polynomial.Polynomial`.  From a symbolic perspective, this bring us closer to the work surrounding formal power series.

Since we don't want to add more NumPy `Op`s for no reason, we also need to consider how prevalent these types of operations are and how well we could identify them through symbolic means.  In other words, having a `Polynomial` `Op` that handles polynomials efficiently (either in terms of evaluation or optimizations) is only useful when it gets used, which means that the relevance of such `Op`s can be directly related to our ability to automatically identify polynomials when the `Op`s aren't explicitly used.

Note that such polynomial identification would need to occur in the context of `Elemwise` operations, since that's what seems to make up the majority of scalar operations.  For instance, `y` in the simple case of
```python
import aesara.tensor as aet

x = aet.vector("x")
a, b, c = aet.scalars("abc")

y = a * x**2 + b * x + c
```
would need to be identified as a polynomial in `x` and converted to a `Polynomial(x, [a, b, c])` node.  

Now, that particular case is easy, but I'm guessing that it's also not the most common.  Perhaps cases like that appear as `c + x * (b + a * x)` more often.  That case isn't too difficult to pick up either, but we would need to cover both (and more) cases elegantly in a single set of concise rewrites/transformations.

**Comment by ricardoV94** [comment]
There are also the cases where some of the scalars may be zero / missing which makes identification perhaps more difficult, but the number of user cases more frequent.

In terms of general use, I would think polynomial regression (up to 2, 3 degrees) is not that uncommon.

As per your advice I will first see if I can come up with a parsimonious set of matching rules.

**Comment by ricardoV94** [comment]
Here is what I come up with. I am not sure how feasible this is in aesara:

Three broad rules:
1. `scalar * x**n = Polyval(x, [0]*n + [scalar])` - This creates the base polyval Ops, which are useless by default.
2. `Polyval(x, c1) + Polyval(x, c2) = Polyval(x, c1 + c2)`
2.1. Special case: `scalar + Polyval(x, c) = Polyval(x, [c[0] + scalar, c[1], ..., c[n]])`
2.2. Special case: `x + Polyval(x, c) = Polyval(x, [c[0], c[1] + 1, c[2], ..., c[n]])`
2.3. Special case: `scalar * x + Polyval(x, c) = Polyval(x, [c[0], c[1] + scalar, c[2], ..., c[n]])`
3. `Polyval(x, c1) * Polyval(x, c2) = Polyval(x, shifted_prod_sum(c1, c2)`
3.1. Special case: `scalar * Polyval(x, c) = Polyval(scalar * c)`
3.2. Special case: `x * Polyval(x, c) = Poylval(x, [0] + [c])`
3.3. Special case: `scalar * x * Polyval(x, c) = Polyval(x, scalar * ([0] + [c]))`

Where `[a] + [b]` represents the concatenation of a with b

Two selective rules, which we apply after the recursive use of rules 1-3 has come to a terminal state:
4. `Polyval(x, c) * Polyval(y, c) = Polyval(x*y, c)`, when `c` are the same.
5. Destroy ulesess sparse Polyval ops, e.g., `Polyval(x, [0, 0, 0, 1]) = x**4`

The `shifted_prod_sum` produces the right `c` vector when multiplying Polyvals. Probably there is a numpy vectorized function that achieves this. The nonvectorized version logic is something like this:

```python
def shifted_prod_sum(c1, c2):
  ret = np.zeros(len(c1) + len(c2) - 1)
  for shift1, c1_ in enumerate(c1):
    for shift2, c2_ in enumerate(c2):
      ret[shift1 + shift2] += c1_ * c2_
return ret
```

In the two examples above these would be the rewrite steps:
```
a * x**2 + b * x + c
> Polyval(x, [0, 0, a]) + Polyval(x, [0, b]) + c  # rule 1, twice
> Polyval(x, [0, b, a]) + c  # rule 2
> Polyval(x, [c, b, a])  # rule 2.1

c + x * (b + a * x)
> c + Polyval(x, [0, 1]) * (b + Polyval(x, [0, a]))  # rule 1, twice
> c + Polyval(x, [0, 1]) * Polyval(x, [b, a]) # rule 2
> c + Polyval(x, [0, b, a])  # rule 3
> Polyval(x, [c, b, a])  # rule 2.1
```
We can make a trade-off between discoverability and rewrite speed by deciding on the minimum exponent needed to attempt the first polyval rewrites (rule 1). Let's say that we only apply rule 1 to `scalar + x**n` if `n >= 2`. Then the first example would still work, but not the second:

```
a * x**2 + b * x + c
> Polyval(x, [0, 0, a]) + b * x + c  # rule 1, once
> Polyval(x, [0, b, a]) + c  # rule 2.3
> Polyval(x, [c, b, a])  # rule 2.1

c + x * (b + a * x)
> c + c + x * (b + a * x)  # we never apply rule 1, and so we never find out we can get a simple Polyval at the end
```

Does this make sense?

**Comment by brandonwillard** [comment]
Yes, the polynomial algebra should be implemented&mdash;at least like it is in NumPy&mdash;and that could take care of a lot.  The difficulties start to appear when it's not entirely clear that a polynomial should be created in `x` instead of, say, another non`Constant` vector `y`.

Also, I came across [this related article recently](https://www.texmacs.org/joris/stablemult/stablemult.html).
